### ISSUE TITLE
Infinite POST request loop

### ISSUE THREAD

**Issue by tpkahlon** [t49tran/react-google-recaptcha-v3#50]
Environment: NEXT.js

```js
// Recaptcha component
<GoogleReCaptchaProvider reCaptchaKey={process.env.RECAPTCHA_KEY}>
  <GoogleReCaptcha
    onVerify={token => {
      onVerifyCaptcha(token);
    }}
  />
</GoogleReCaptchaProvider>
```

```js
// pages/access/index.js
const [loaded, setLoaded] = useState(false);
const onVerifyCaptcha =(token) => {
  setLoaded(true); // triggers infinite POST request loop
};
<Recaptcha onVerifyCaptcha={onVerifyCaptcha} />
```

Possible Solution:

```js
// pages/access/index.js
let validToken = '';

useEffect(() => {
        setLoaded(true);
}, [loaded]);
    
const onVerifyCaptcha = useCallback((token) => {
        validToken = token;
}, [validToken]);
```

Any feedback/advice regarding this solution or any possible alternative is appreciated. How to use HOC, Class based component with memoizedCallback in order to prevent infinite loop bug?

**Comment by DanLeCornu** [comment]
To clarify - are you saying that when using React's useCallback to memoize the function, that the infinite loop issue stops? Because I read in the docs that it should stop that - however, even when I'm memoizing my onVerify function, it's still infinite looping:
```
  const onVerify = React.useCallback(
    // INFINITE LOOP HAPPENING HERE STILL
    (token: string) => {
       // DOING STUFF
    },
    [],
  )

  return (
    <GoogleReCaptchaProvider useRecaptchaNet reCaptchaKey={RECAPTCHA_KEY}>
      <GoogleReCaptcha action="advisorRegistration" onVerify={onVerify} />
    </GoogleReCaptchaProvider>
  )
```

Perhaps it's the way I'm using the provider? @t49tran could you shed any light on this?

**Comment by t49tran** [comment]
It depends on what you do inside your `onVerify` function.

Clearly, not memoizing the function and at the same time, update the state will lead to infinite loop.

If your `onVerify` function is just something like:

```javascript
const onVerify = React.useCallback(
    // INFINITE LOOP HAPPENING HERE STILL
    (token: string) => {
       // DOING STUFF
    },
    [],
  )
```

I couldn't see how an infinite loop can be triggered.

You guys can check the example to see how to use the `onVerify` function without leading to an infinite loop.

**Comment by t49tran** [comment]
> To clarify - are you saying that when using React's useCallback to memoize the function, that the infinite loop issue stops? Because I read in the docs that it should stop that - however, even when I'm memoizing my onVerify function, it's still infinite looping:
> 
> ```
>   const onVerify = React.useCallback(
>     // INFINITE LOOP HAPPENING HERE STILL
>     (token: string) => {
>        // DOING STUFF
>     },
>     [],
>   )
> 
>   return (
>     <GoogleReCaptchaProvider useRecaptchaNet reCaptchaKey={RECAPTCHA_KEY}>
>       <GoogleReCaptcha action="advisorRegistration" onVerify={onVerify} />
>     </GoogleReCaptchaProvider>
>   )
> ```
> 
> Perhaps it's the way I'm using the provider? @t49tran could you shed any light on this?

You should never put the provider `<GoogleReCaptchaProvider />` in the same component with `<GoogleReCaptcha />`, doing it will make the Provider to be re-rendered whenever changes happened in the child component, re-creating the context and lead to issues. 

A lot of these infinite rendering errors come down to understanding how React reconciliation and rendering works. It will take more than a few lines to explain it and it's not the scope of this project.

In summary, the `GoogleReCaptchaProvider` is to provide a React context that contains recaptcha function to the rest of the React tree, it should be included only one per tree, and place as high to the top of the tree as possible. Whenever this component is re-rendered, it will provide a new value to the context.

It's actually not something rocket-science and if you have used a popular React libraries such as Redux or Material UI, you should have seen these providers component. Generally speaking, they act like a singleton.

I will close this issue as it's not a feature or error that requires action.

**Comment by tpkahlon** [comment]
Okay, so here is what I found:

a. In case of useCallback hook, when you perform a setState in onVerify method in your page component in Next, it runs into making infinite POST calls. It only happens if you setState as object. If it is like setState(false) or setState('foo'), it does not make infinite POST calls.

**Solution:** I solved it by using useMemo function. See the demo [here](https://github.com/tpkahlon/react-google-recaptcha-v3).

Please share your thoughts on the solution. If you feel it is good, you can close the issue.

**Comment by t49tran** [comment]
This has been documented in the ReadMe.md.

**Comment by DanLeCornu** [comment]
@t49tran You are correct - I just needed to wrap the provider outside of the component, and we're all good. Thanks for the insight, and the quick responses!

**Comment by chtpl** [comment]
Does including the provider send data to Google about the client already? Or is this only done once a 
`<GoogleReCaptcha onVerify={handleVerify} />` is rendered? 

I plan to only activate the captcha until my users enable the "I accept the privacy policy" checkbox on the registration page. I would do that with a conditional rendering. Is this the correct way to do it?
